Adds `Error.cause` polyfill

mhassan1 opened the pull request:
This PR adds a polyfill for `Error.cause`.

JakeChampion commented:
/ok-to-test sha=a883575bbd5

romainmenke commented:
/ok-to-test sha=a12066d

romainmenke commented on polyfills/Error/cause/config.toml:
```diff
@@ -0,0 +1,28 @@
+firefox_mob = "<91"
+ie = "*"
+ie_mob = "*"
+opera = "*"
+op_mob = "*"
+op_mini = "*"
+safari = "<15"
+ios_saf = "<15"
```
```suggestion
opera = "<79"
op_mob = "<66"
op_mini = "*"
safari = "<15.0"
ios_saf = "<15.0"
```

romainmenke reviewed (commented):
MDN data for Opera seems incorrect, seems supported from 79

<img width="345" alt="Screenshot 2022-09-10 at 11 49 59" src="https://user-images.githubusercontent.com/11521496/189478090-33a3cb37-f25c-4fd0-bbec-7a2ec89c547d.png">

also reported upstream : https://github.com/mdn/browser-compat-data/issues/17750

zloirock commented on polyfills/Error/cause/polyfill.js:
```diff
@@ -0,0 +1,72 @@
+		var name = _errorNames[i];
+		_nativeErrors[name] = self[name];
+		_errorConstructors[name] = _makeErrorConstructor(name, _newErrors[name]);
+		CreateMethodProperty(self, name, _inheritErrorPrototype(name, _newErrors[name]));
```
`Error.__proto__ === Function.prototype` should be `true`, but you inherit even base `Error`.

mhassan1 commented:
@romainmenke This should be ready for re-run. Thanks.

JakeChampion commented:
/ok-to-test sha=7e4a75c

mhassan1 commented:
@JakeChampion I've fixed a test for IE9 and IE10. This is ready for re-run.

JakeChampion commented:
/ok-to-test sha=7f9adf1

mhassan1 commented:
All tests passed. This is ready for review.

JakeChampion commented on polyfills/Error/cause/polyfill.js:
```diff
@@ -0,0 +1,80 @@
+	CreateMethodProperty(self, 'TypeError',      _inheritErrorPrototype(_newErrors.TypeError));
+	CreateMethodProperty(self, 'URIError',       _inheritErrorPrototype(_newErrors.URIError));
+
+	if ('AggregateError' in self) {
```
We can define polyfill dependies in config.toml, should we be adding one of these as a dependency of the other?

mhassan1 replied:
Technically, neither one should be marked as a dependency of the other, since someone may want one but not both. Maybe we need a soft dependency concept, where we ensure the order only when both are requested.

Alternatively, I can externalize this polyfill's logic and apply it inside `AggregateError`, when necessary. Or, I can add another polyfill for `AggregateError.cause`.

mhassan1 replied:
I just pushed an implementation of the second option:

> externalize this polyfill's logic and apply it inside `AggregateError`, when necessary

Let me know what you think.

mhassan1 commented:
@JakeChampion Apologies, this needs another review. I refactored the `Error.cause` logic into an internal `_InstallErrorCause` polyfill, which we can use in both `Error.cause` and `AggregateError` (and any other error types that come up in the future). I also added a couple related fixes (and tests) to `AggregateError`.

JakeChampion commented:
/ok-to-test sha=e6bdf95

mhassan1 commented:
@JakeChampion I see a couple flaky test failures.

JakeChampion commented:
> @JakeChampion I see a couple flaky test failures.

rerunning now 👍
